src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,422 @@
+	 */
+	appendStringInfo(CommonValueQuery,
+					 "SELECT shardid , unnest(result::%s[]) from run_command_on_shards(%s,$$SELECT array_agg(val)"
+					 " FROM pg_stats s , unnest(most_common_vals::text::%s[],most_common_freqs) as res(val,freq)"
+					 " WHERE tablename = %s AND attname = %s AND schemaname = %s AND freq > %lf $$)"
+					 " WHERE result <> '' AND shardid = %ld;",
+					 dataType, quote_literal_cstr(tableName),
+					 dataType,
+					 quote_literal_cstr(shardName),
+					 quote_literal_cstr(distributionColumnName),
+					 quote_literal_cstr(get_namespace_name(get_rel_namespace(
+															   tableId))),
+					 TenantFrequency,
+					 shardId);
+
+	ereport(DEBUG4, errmsg("%s", CommonValueQuery->data));
+	List *splitPoints = NULL;
+
+	/* Saving the current memory context*/
+	MemoryContext originalContext = CurrentMemoryContext;
+
+	SPI_connect();
+	SPI_exec(CommonValueQuery->data, 0);
+
+	/*Saving the SPI memory context for switching*/
+	MemoryContext spiContext = CurrentMemoryContext;
+
+	int64 rowCount = SPI_processed;
+	int32 hashedValue;
+
+	ereport(DEBUG4, errmsg("%ld", rowCount));
+
+	if (rowCount > 0)
+	{
+		/*For every common tenant value split point is calculated on the basis of
+		 * the hashed value and the unique split points are appended to the list
+		 * and the resulting is then sorted and returned.
+		 */
+		SPITupleTable *tupletable = SPI_tuptable;
+		CitusTableCacheEntry *cacheEntry = GetCitusTableCacheEntry(tableId);
+		for (int rowIndex = 0; rowIndex < rowCount; rowIndex++)
+		{
+			HeapTuple tuple = tupletable->vals[rowIndex];
+			char *commonValue = SPI_getvalue(tuple, tupletable->tupdesc, 2);
+			ereport(DEBUG4, errmsg("%s", commonValue));
+			Datum tenantIdDatum = StringToDatum(commonValue,
+												ColumnTypeIdForRelationColumnName(tableId,
+																				  distributionColumnName));
+			Datum hashedValueDatum = FunctionCall1Coll(cacheEntry->hashFunction,
+													   cacheEntry->partitionColumn->
+													   varcollid,
+													   tenantIdDatum);
+			hashedValue = DatumGetInt32(hashedValueDatum);
+			ereport(DEBUG4, errmsg("%d", hashedValue));
+
+			/*Switching the memory context to store the unique SplitPoints in a list*/
+
+			MemoryContextSwitchTo(originalContext);
+			if (hashedValue == shardMinValue)
+			{
+				splitPoints = list_append_unique_int(splitPoints, hashedValue);
+			}
+			else if (hashedValue == shardMaxValue)
+			{
+				splitPoints = list_append_unique_int(splitPoints, hashedValue - 1);
+			}
+			else
+			{
+				splitPoints = list_append_unique_int(splitPoints, hashedValue - 1);
+				splitPoints = list_append_unique_int(splitPoints, hashedValue);
+			}
+			MemoryContextSwitchTo(spiContext);
+		}
+		SPI_freetuptable(tupletable);
+		list_sort(splitPoints, list_int_cmp);
+	}
+	else
+	{
+		StringInfo AvgHashQuery = makeStringInfo();
+		uint64 tableSize = 0;
+		bool check = DistributedTableSize(tableId, TOTAL_RELATION_SIZE, true,
```

Comment:
since our query below only really reads from a single shard, I'm kind of wondering whether we could just reuse shardSize (not shardGroupSize) here.

Reply:
using shardSize now